### PR TITLE
Stochastic mech features

### PR DESCRIPTION
**Pull request by AurelienJaquier**
- new method that either set to deterministic, or set back to stochastic (if stochastic mechanism)
- always set stoch data, even when used in deterministic mode. Reason: some neuron function (e.g. psection) will look up the mechanism data, and if it points to nothing, it crashes

### CODE REVIEW

**Comment by DrTaDa** [comment]
@wvangeit this is something that was implemented for BluePyEModel but we thought it might be better in BluePyOpt

File: bluepyopt/ephys/mechanisms.py [hunk 1]
```diff
@@ -85,6 +85,9 @@ def __init__(
         self.preloaded = preloaded
         self.cell_model = None
         self.deterministic = deterministic
+        # If deterministic is False, it might get changed to True.
```
Comment (wvangeit) [inline bluepyopt/ephys/mechanisms.py:88]: I think we need a bit better description here. It's a bit weird code, so I think the use case should be documented well

File: bluepyopt/ephys/mechanisms.py [hunk 2]
```diff
@@ -143,6 +147,13 @@ def instantiate_determinism(self, deterministic, icell, isec, sim):
                     'Stoch channel, not %s' %
                     self.suffix)
 
+    def set_stochasticity(self, allow_stochasticity):
+        """Set stochasticity if allowed, else force determinism"""
+        if allow_stochasticity and not self.always_deterministic:
+            self.deterministic = False
+        else:
+            self.deterministic = True
```
Comment (wvangeit) [inline bluepyopt/ephys/mechanisms.py:155]: Would there be a possibility of raising an exception here. If I call set_stochastic on something, I assume the object will be stochastic afterwards, unless if it throws an exception.

Comment (AurelienJaquier) [reply bluepyopt/ephys/mechanisms.py:155]: The function is more of 'whether this should be stochastic or not', than 'set this to stochastic'. The idea is that, for the pre-protocols, we want everything to be deterministic, so we use this method with allow_stochasticity=False, which sets everything to deterministic. Once the preprotocols are done, we want to set back the stochasticity of each mechanism to its original state (True/False), and so we call this function with allow_stochasticity=True. Note that even when stochasticity is allowed, not every mech will be set to non-deterministic.

File: bluepyopt/ephys/mechanisms.py (hunk 1 above)
Comment (AurelienJaquier) [reply bluepyopt/ephys/mechanisms.py:88]: 'One might want to temporarily set a stochastic mechanism to deterministic to run a few protocols, and then turn it back to stochastic mode to run some other protocols. The always_deterministic attribute remembers the mechanisms that were set to be deterministic or stochastic when created. If one wants to turn the stochasticity back on, after temporarily forcing the mechanisms to be deterministic, one can set the deterministic attribute to the value of always_deterministic using the self.set_stochasticity(True) method'

Would a comment like this be acceptable?

File: bluepyopt/ephys/mechanisms.py (hunk 2 above)
Comment (DrTaDa) [reply bluepyopt/ephys/mechanisms.py:155]: (pre-protocols are protocols such as the ones to search for RMP, Rin, rheobase ..)

**Comment by DrTaDa** [comment]
@AurelienJaquier it makes me think, maybe all of the stochastic logic could be brought to BluePyOpt. We should have a look.

**Comment by AurelienJaquier** [comment]
@DrTaDa Do you mean the logic in the run() method of BPEM_Protocol? Or were you thinking of something else?

**Comment by wvangeit** [comment]
Should the pre_protocol change the value to deterministic, run it, and set it back to the old value?

**Comment by DrTaDa** [comment]
> @DrTaDa Do you mean the logic in the run() method of BPEM_Protocol? Or were you thinking of something else?

@AurelienJaquier yes, this part.

**Comment by AurelienJaquier** [comment]
Ok, I'll try both your ideas.

**Comment by DrTaDa** [comment]
Looks good to me

**Review by wvangeit — APPROVED** [review]

**Pull request merged**